src/main.cpp

```diff
@@ -4835,10 +4894,39 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                     CNodeState *nodestate = State(pfrom->GetId());
                     if (CanDirectFetch(chainparams.GetConsensus()) &&
                         nodestate->nBlocksInFlight < MAX_BLOCKS_IN_TRANSIT_PER_PEER) {
-                        vToFetch.push_back(inv);
+                        if (nodestate->fProvidesHeaderAndIDs)
+                            vToFetch.push_back(CInv(MSG_CMPCT_BLOCK, inv.hash));
+                        else
+                            vToFetch.push_back(inv);
                         // Mark block as in flight already, even though the actual "getdata" message only goes out
                         // later (within the same cs_main lock, though).
                         MarkBlockAsInFlight(pfrom->GetId(), inv.hash, chainparams.GetConsensus());
+                        if (mapBlocksInFlight.size() <= 2) {
+                            // We seem to be rather well-synced, so it appears pfrom was the first to provide us
+                            // with this block! Let's get them to announce using compact blocks in the future.
+                            if (nodestate->fProvidesHeaderAndIDs) {
+                                bool fAlreadyAnnouncing = false;
+                                BOOST_FOREACH(const NodeId nodeid, lNodesAnnouncingHeaderAndIDs)
+                                    if (nodeid == pfrom->GetId())
+                                        fAlreadyAnnouncing = true;
+                                if (!fAlreadyAnnouncing) {
+                                    bool fAnnounceUsingCMPCTBLOCK = false;
+                                    uint64_t nCMPCTBLOCKVersion = 1;
+                                    if (lNodesAnnouncingHeaderAndIDs.size() >= 3) {
```

Comment:
nit: constant will be nice, 3 is reused on https://github.com/bitcoin/bitcoin/pull/8068/files#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR5499